cpp/examples/nested_types/deduplication.cpp

```diff
@@ -0,0 +1,205 @@
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/groupby.hpp>
+#include <cudf/io/json.hpp>
+#include <cudf/io/types.hpp>
+#include <cudf/join.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/mr/device/cuda_memory_resource.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/owning_wrapper.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+#include <chrono>
+#include <iostream>
+#include <string>
+
+/**
+ * @file deduplication.cpp
+ * @brief Demonstrates usage of the libcudf APIs to perform operations on nested-type tables.
+ *
+ * The algorithms chosen to be demonstrated are to showcase nested-type row operators of three
+ * kinds:
+ * 1. hashing: Used by functions `count_aggregate` and `join_count` to hash inputs of any type
+ * 2. equality: Used by functions `count_aggregate` and `join_count` in conjunction with hashing
+ *              to determine equality for nested types
+ * 3. lexicographic: Used by function `sort_keys` to create a lexicographical order for nested-types
+ * so as to enable sorting
+ *
+ * @note This example is for demonstration purposes only. It is not intended to show the most
+ * performant way to do the example algorithm.
+ *
+ */
+
+/**
+ * @brief Create CUDA memory resource
+ */
+auto make_cuda_mr() { return std::make_shared<rmm::mr::cuda_memory_resource>(); }
+
+/**
+ * @brief Create a pool device memory resource
+ */
+auto make_pool_mr()
+{
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(make_cuda_mr());
+}
+
+/**
+ * @brief Create memory resource for libcudf functions
+ */
+std::shared_ptr<rmm::mr::device_memory_resource> create_memory_resource(std::string_view name)
+{
+  if (name == "pool") { return make_pool_mr(); }
+  return make_cuda_mr();
+}
+
+cudf::io::table_with_metadata read_json(std::string filepath)
+{
+  auto source_info = cudf::io::source_info(filepath);
+  auto builder     = cudf::io::json_reader_options::builder(source_info).lines(true);
+  auto options     = builder.build();
+  return cudf::io::read_json(options);
+}
+
+void write_json(cudf::table_view tbl, cudf::io::table_metadata metadata, std::string filepath)
+{
+  // write the data for inspection
+  auto sink_info = cudf::io::sink_info(filepath);
+  auto builder   = cudf::io::json_writer_options::builder(sink_info, tbl).lines(true);
+  builder.metadata(metadata);
+  auto options = builder.build();
+  cudf::io::write_json(options);
+}
+
+/**
+ * @brief Aggregate count of duplicate rows in nested-type column
+ */
+std::unique_ptr<cudf::table> count_aggregate(cudf::table_view tbl)
+{
+  // Get count for each key
+  auto keys = cudf::table_view{{tbl.column(0)}};
+  auto val  = cudf::make_numeric_column(cudf::data_type{cudf::type_id::INT32}, keys.num_rows());
```

Comment:
Do count aggregations return `int32_t` or `size_type`? I think it's supposed to be `size_type` but not 100% sure.
```suggestion
  auto val  = cudf::make_numeric_column(cudf::data_type{cudf::type_to_id<cudf::size_type>()}, keys.num_rows());
```

Reply:
It's just a dummy column to force usage of hash aggregation instead of sort aggregation, it does not hold the return values.

Reply:
I'm not sure if I understand. We usually want to rely on hash aggregation anytime we can, since it tends to be faster than sort aggregations. Can you point to the behavior that you're avoiding, and why this is the right solution (as opposed to fixing the aggregation dispatch to use hashing instead of sorting)?

Reply:
https://github.com/rapidsai/cudf/blob/abc0d41d1d9033d581948ae19384e0aa0f33da77/cpp/src/groupby/hash/groupby.cu#L654-L656

Reply:
This isn't the right solution. It should definitely be fixed at the core of the problem.

Reply:
Is this something we now have the machinery to fix? I'm not sure what stops this from working. We should have full nested type support now.

Reply:
If we don't force hash aggregations do we still have at least some? As long as the example is demonstrating the all the comparator functionality that we want I think it's totally fine if it isn't the most performant right now (and we've removed timings anyway). If there are improvements to be made to libcudf, we should open a separate issue for that.

Reply:
I don't get it, what do you mean by "at least have some"?

Reply:
1. We need to file an issue that makes this use hash-based aggregations without any "forcing"
2. I think Vyas is asking, if you don't add this column to "force" hash aggregations, are _some_ of the aggregations hash-based and others sort-based? My understanding of libcudf's behavior is that if any aggregation is sort-based, all the aggregations fall back to using sort-based implementations. Is that true?

Reply:
Yes, that's right. If an aggregation only has a sort-based implementation then it will fall back to a sort aggregation.

In this particular case though, a hash-based aggregation is available and it needs to be "forced" due to a limitation.

Here's the issue: https://github.com/rapidsai/cudf/issues/14412

Reply:
@bdice are you OK with marking this thread as resolved for now? The issue looks like it has a reasonable plan for how to move forward, but I don't think we should block this PR on getting that enabled.